lib/collection/src/collection_manager/segments_searcher.rs

```diff
@@ -629,4 +634,32 @@ mod tests {
     fn test_sampling_limit_high() {
         assert_eq!(sampling_limit(1000000, None, 464530, 35103551), 1000000);
     }
+
+    /// Tests whether calculating the effective ef limit value is correct.
+    ///
+    /// Because there was confustion about what the effective value should be for some imput
+    /// combinations, we decided to write this tests to ensure correctness.
+    ///
+    /// See: <https://github.com/qdrant/qdrant/pull/1694>
+    #[test]
+    fn test_effective_limit() {
+        // Test cases to assert: (limit, ef_limit, poisson_sampling, effective)
+        let tests = [
+            (1000, 128, 150, 150),
+            (1000, 128, 110, 128),
+            (130, 128, 150, 130),
+            (130, 128, 110, 128),
+            (50, 128, 150, 50),
+            (50, 128, 110, 50),
+            (500, 1000, 300, 500),
```

Comment:
This assumes that with

- limit: 500
- ef_limit: 1000
- poisson_sampling: 300

the effective limit **must be 500** (not 1000).

Is that correct?

Reply:
it should be 1000

Reply:
Great! That means the current if-statement approach has problems. I'll update it accordingly.